src/Microsoft.Data.Analysis.IO/DataFrameIO.cs

```diff
@@ -0,0 +1,266 @@
+                    throw new FormatException(string.Format(Strings.LessColumnsThatExpected, nbline + 1));
+
+                string val = line[col];
+
+                if (string.Equals(val, "null", StringComparison.OrdinalIgnoreCase))
+                {
+                    continue;
+                }
+
+                bool boolParse = bool.TryParse(val, out bool boolResult);
+                if (boolParse)
+                {
+                    res = DetermineType(nbline == 0, typeof(bool), res);
+                    ++nbline;
+                    continue;
+                }
+                else
+                {
+                    if (string.IsNullOrEmpty(val))
+                    {
+                        res = DetermineType(nbline == 0, typeof(bool), res);
+                        continue;
+                    }
+                }
+                bool floatParse = float.TryParse(val, out float floatResult);
+                if (floatParse)
+                {
+                    res = DetermineType(nbline == 0, typeof(float), res);
+                    ++nbline;
+                    continue;
+                }
+                else
+                {
+                    if (string.IsNullOrEmpty(val))
+                    {
+                        res = DetermineType(nbline == 0, typeof(float), res);
+                        continue;
+                    }
+                }
+                res = DetermineType(nbline == 0, typeof(string), res);
+                ++nbline;
+            }
+            return res;
+        }
+
+        public static DataFrame LoadCsv(Func<Stream> csvStream,
+                                Action<TextFieldParser> SetTextFieldParserProperties,
+                                char separator = ',', bool header = true,
+                                string[] columnNames = null, Type[] dataTypes = null,
+                                long numberOfRowsToRead = -1, int guessRows = 10, bool addIndexColumn = false,
+                                Encoding encoding = null)
+        {
+            Stream stream = csvStream();
+            if (!stream.CanSeek)
+            {
+                throw new ArgumentException(Strings.NonSeekableStream, nameof(stream));
+            }
+
+            if (dataTypes == null && guessRows <= 0)
+            {
+                throw new ArgumentException(string.Format(Strings.ExpectedEitherGuessRowsOrDataTypes, nameof(guessRows), nameof(dataTypes)));
+            }
+
+            var linesForGuessType = new List<string[]>();
+            long rowline = 0;
+            int numberOfColumns = dataTypes?.Length ?? 0;
+
+            if (header == true && numberOfRowsToRead != -1)
+            {
+                numberOfRowsToRead++;
+            }
+
+            List<DataFrameColumn> columns;
+            // First pass: schema and number of rows.
+            using (TextFieldParser parser = new TextFieldParser(stream, defaultEncoding: encoding ?? Encoding.UTF8))
```

Comment:
Uses TextFieldParser instead of reading a stream. TextFieldParser handles quotes correctly and is in-box